src/mockgyver.erl

```diff
@@ -1372,73 +1372,103 @@ ensure_renamed_mod_to_load_aux(RenamedMod, #modinfo{code=Code}) ->
-    %%             ActionFun ->
-    %%                 ActionFun(A2, A1)
-    %%         end.
-    Args = mk_args(A),
-    Body =[erl_syntax:case_expr(
-             mk_call(mockgyver, reg_call_and_get_action,
-                     [erl_syntax:tuple([erl_syntax:abstract(Mod),
-                                        erl_syntax:abstract(F),
-                                        erl_syntax:list(Args)])]),
-             [erl_syntax:clause([erl_syntax:atom(undefined)],
-                                none,
-                                [mk_call(erlang, error,
-                                         [erl_syntax:atom(undef)])]),
-              erl_syntax:clause([erl_syntax:variable('ActionFun')],
-                                none,
-                                [mk_call('ActionFun', Args)])])],
-    erl_syntax:function(
-      erl_syntax:abstract(F),
-      [erl_syntax:clause(Args, none, Body)]).
+    %% It is many times faster than constructing a number of functions,
+    %% each containing the inner case expression, though.
+    func_from_str_fmt(
+      "'$handle_undefined_function'(FnName, Args) ->
+           Arity = length(Args),
+           case lists:member({FnName, Arity}, ~p) of
+               true ->
+                   case mockgyver:reg_call_and_get_action({~p,FnName,Args}) of
+                       undefined ->
+                           ~s;
+                       ActionFun ->
+                          try
+                              apply(ActionFun, Args)
+                          catch
+                              Class:Reason:St0 ->
+                                  ToMF = {~p, FnName},
+                                  St = rewrite_st(ActionFun, ToMF, Arity, St0),
+                                  erlang:raise(Class, Reason, St)
+                          end
+                   end;
+               false ->
+                   error_handler:raise_undef_exception(~p, FnName, Args)
+           end.",
+      [ExportedFAs, Mod, FmtNoAction("FnName", "Args"), Mod, Mod]).
+
+%% Ensure a stacktrace like the following contains the mocked module
+%% and function, instead of an internal fun that the user was not
+%% involved in creating.
+%%
+%% Test code:
+%%
+%%     ?WHEN(mockgyver_dummy:return_arg(N) -> error(foo)),
+%%     mockgyver_dummy:return_arg(1),
+%%
+%% Before:
+%%
+%%      Failure/Error: {error,function_clause,
+%%                         [{mockgyver_tests,
+%%                              '-some_test/1-fun-0-',
+%%                              [-1],
+%%                              [...]},
+%%                          ...
+%%
+%% After:
+%%
+%%      Failure/Error: {error,function_clause,
+%%                         [{mockgyver_dummy,
+%%                              return_arg,
+%%                              [-1],
+%%                              [...]},
+%%                          ...
+mk_rewrite_st_function(Mod) ->
+    func_from_str_fmt(
+      "rewrite_st(ActionFun, ToMF, Arity, Stacktrace)
```

Comment:
Do we need to figure out some odd-looking name here? I'm thinking if the module we mock happens to export `rewrite_st` (unlikely, but...) then we would end up here instead of in the `'$handle_undefined_function'/2`. Or alternatively (maybe better?) to skip the `export_all` down in `mk_mod/2` and export only `'$handle_undefined_function'/2`.

Reply:
I named it '$mockgyver_tweak_st' instead. With "rewrite" I got too long lines :-/.
Perhaps I should inline it instead.